capture_metrics_test.go

```diff
@@ -57,29 +62,49 @@ func TestCaptureMetrics(t *testing.T) {
 			WantCode:    http.StatusOK,
 		},
 		{
+			Name: "empty panic",
 			Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
 				panic("oh no")
 			}),
-			WantErr: "EOF",
+			WantCode:     http.StatusOK, // code is not written so is our default
+			WantDuration: 1,             // confirm non-zero
+			WantErr:      "EOF",
+		},
+		{
+			Name: "panic after partial response",
+			Handler: http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+				w.WriteHeader(http.StatusInternalServerError)
+				w.Write([]byte("failed to execute"))
+				panic("oh no")
+			}),
+			WantCode:     http.StatusInternalServerError,
+			WantDuration: 1, // confirm non-zero
+			WantWritten:  17,
+			WantErr:      "EOF",
 		},
 	}
 
 	for i, test := range tests {
-		func() {
+		t.Run(test.Name, func(t *testing.T) {
 			ch := make(chan Metrics, 1)
 			h := http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-				ch <- CaptureMetrics(test.Handler, w, r)
+				m := Metrics{Code: http.StatusOK}
+				defer func() {
+					ch <- m
+				}()
+				m.CaptureMetrics(w, func(ww http.ResponseWriter) {
+					test.Handler.ServeHTTP(ww, r)
+				})
 			})
 			s := httptest.NewServer(h)
 			defer s.Close()
 			res, err := http.Get(s.URL)
 			if !errContains(err, test.WantErr) {
 				t.Errorf("test %d: got=%s want=%s", i, err, test.WantErr)
 			}
-			if err != nil {
-				return
+			if res != nil && res.Body != nil {
+				defer res.Body.Close()
 			}
```

Comment:
```suggestion
			if err != nil {
				// continue on error
			} else {
				defer res.Body.Close()
 			}
```

Because, I'm unsure this would be more readable

```suggestion
			if err == nil {
				defer res.Body.Close()
 			}
```

Reply:
Thanks - second one seemed clearer to me